fuel/datasets/cifar10.py

```diff
@@ -40,46 +33,31 @@ class CIFAR10(IndexableDataset):
 
     """
     provides_sources = ('features', 'targets')
-    folder = 'cifar10'
-    files = {
-        'train': [os.path.join('cifar-10-batches-py',
-                               'data_batch_{}'.format(i))
-                  for i in range(1, 6)],
-        'test': ['cifar-10-batches-py/test_batch']
-    }
+    filename = 'cifar10.hdf5'
 
-    def __init__(self, which_set, flatten=True, **kwargs):
+    def __init__(self, which_set, flatten=True, dtype=config.floatX,
+                 **kwargs):
         if which_set not in ('train', 'test'):
             raise ValueError("CIFAR10 only has a train and test set")
 
+        kwargs.setdefault('load_in_memory', True)
         self.which_set = which_set
-        self.flatten = flatten
+        self.reshape = flatten
+        self.dtype = dtype
 
-        super(CIFAR10, self).__init__(OrderedDict(zip(self.provides_sources,
-                                                      self._load_cifar10())),
+        super(CIFAR10, self).__init__(self.data_path, which_set,
                                       **kwargs)
 
-    def load(self):
-        self.indexables = [data[self.start:self.stop] for source, data
-                           in zip(self.provides_sources, self._load_cifar10())
-                           if source in self.sources]
-
-    def _load_cifar10(self):
-        base_path = os.path.join(config.data_path, self.folder)
-        num_examples = 50000 if self.which_set == 'train' else 10000
-        image_shape = (3072,) if self.flatten else (3, 32, 32)
-        images = numpy.zeros((num_examples,) + image_shape,
-                             dtype=config.floatX)
-        labels = numpy.zeros((num_examples, 1), dtype='uint8')
-        for i, fname in enumerate(self.files[self.which_set]):
-            with open(os.path.join(base_path, fname), 'rb') as f:
-                if six.PY3:
-                    batch = cPickle.load(f, encoding='latin1')
-                else:
-                    batch = cPickle.load(f)
-                if not self.flatten:
-                    batch['data'] = batch['data'].reshape((10000, 3, 32, 32))
-                images[10000 * i:10000 * (i + 1)] = batch['data']
-                labels[10000 * i:10000 * (i + 1)] = numpy.asarray(
-                    batch['labels'], dtype='uint8')[:, None]
-        return images, labels
+    @property
+    def data_path(self):
+        return os.path.join(config.data_path, self.filename)
+
+    def get_data(self, state=None, request=None):
+        rval = list(super(CIFAR10, self).get_data(state=state,
+                                                  request=request))
+        if 'features' in self.sources:
+            i = self.sources.index('features')
+            rval[i] = rval[i].astype(self.dtype)
```

Comment:
You should remove `self.dtype` from the constructor and instead use `config.floatX` here. Also, please divide by 255.0 before casting.

The two lines below should be removed, as `H5PYDataset` should take care of the `flatten` constructor argument by itself.